include/swift/AST/Attr.h

```diff
@@ -1199,6 +1199,29 @@ class RestatedObjCConformanceAttr : public DeclAttribute {
   }
 };
 
+/// Directly controls the mangled name of this type --- not the entire symbol,
+/// just the name part.
+class CustomTypeNameManglingAttr : public DeclAttribute {
+public:
+  explicit CustomTypeNameManglingAttr(Identifier name,
+                                      char relatedEntityKind = '\0')
+    : DeclAttribute(DAK_CustomTypeNameMangling, SourceLoc(), SourceRange(),
+                    /*Implicit=*/true),
+      Name(name), RelatedEntityKind(relatedEntityKind) {
+    assert(!relatedEntityKind ||
+           (relatedEntityKind >= 'a' && relatedEntityKind <= 'j') ||
+           (relatedEntityKind >= 'A' && relatedEntityKind <= 'J') &&
+           "only [A-Ja-j] are supported as related entity kinds");
```

Comment:
It seems wrong to actually use the mangling in the representation.  Maybe have a Kind and then map to/from the char?

Reply:
I thought about that, but then you just end up with an enum of 20 cases with no intrinsic meaning (because "context-dependent"). I could have restricted it to a number 1-20 or 0-19, but that's less debuggable.

Reply:
Why would it need to be context-dependent?  You can use a semantically meaningful enum and then map it to the context-dependent case.

Reply:
John and I talked about this in person and he convinced me to make something more semantic here, like ClangImporterSynthesizedType.